src/vs/workbench/services/textMate/browser/arrayOperation.ts

```diff
@@ -0,0 +1,95 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { compareBy, numberComparator } from 'vs/base/common/arrays';
+import { ContiguousGrowingArray } from 'vs/editor/common/model/textModelTokens';
+
+export class ArrayEdit {
+	public readonly edits: readonly SingleArrayEdit[];
+
+	constructor(
+		/**
+		 * Disjoint edits that are applied in parallel
+		 */
+		edits: readonly SingleArrayEdit[]
+	) {
+		this.edits = edits.slice().sort(compareBy(c => c.offset, numberComparator));
+	}
+
+	applyTo(array: ContiguousGrowingArray<any>): void {
+		for (let i = this.edits.length - 1; i >= 0; i--) {
+			const c = this.edits[i];
+			array.delete(c.offset, c.length);
+			array.insert(c.offset, c.newLength);
+		}
+	}
+}
+
+export class SingleArrayEdit {
+	constructor(
+		public readonly offset: number,
+		public readonly length: number,
+		public readonly newLength: number,
+	) { }
+
+	toString() {
+		return `[${this.offset}, +${this.length}) -> +${this.newLength}}`;
+	}
+}
+
+export interface IIndexTransformer {
+	transform(index: number): number | undefined;
+}
+
+export class MonotonousIndexTransformer implements IIndexTransformer {
+	public static fromMany(transformations: ArrayEdit[]): IIndexTransformer {
+		// TODO improve performance by combining transformations first
+		const transformers = transformations.map(t => new MonotonousIndexTransformer(t));
+		return new CombinedIndexTransformer(transformers);
+	}
+
+	private idx = 0;
+	private offset = 0;
+
+	constructor(private readonly transformation: ArrayEdit) {
+	}
+
+	/**
+	 * Precondition: index >= previous-value-of(index).
+	 */
+	transform(index: number): number | undefined {
+		let nextChange = this.transformation.edits[this.idx] as SingleArrayEdit | undefined;
+		while (nextChange && nextChange.offset + nextChange.length <= index) {
+			this.offset += nextChange.newLength - nextChange.length;
+			this.idx++;
+			nextChange = this.transformation.edits[this.idx];
+		}
```

Comment:
The edits are sorted (I'm not sure what direction here), so it feels to me that there's an opportunity to add an early break statement somewhere in here.

Reply:
Notice that `this.offset` changes. I don't think an early break here works.